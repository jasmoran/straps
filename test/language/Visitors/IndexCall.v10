#!IndexCall.out
#-

include "src/stdlib/stdlib.v10"
include "src/language/parser/Parser.v10"
include "src/language/Scope.v10"
include "src/language/Type.v10"

include "src/language/visitors/Visitor.v10"
include "src/language/visitors/Printer.v10"
include "src/language/visitors/Constructor.v10"
include "src/language/visitors/Typing.v10"
include "src/language/visitors/Method.v10"
include "src/language/visitors/Definition.v10"
include "src/language/visitors/Resolving.v10"
include "src/language/visitors/IndexCall.v10"

global(Scope scope, Node root_node)

func void main()(Stream input, Parser parser, Node root, Visitor v, Scope sc) {
	# Open input file
	input = File_open("test/language/Visitors/sample.v10.test")
	if (input < 0) {
		"Couldn't read input file".println(STDERR)
		return 1
	}

	parser = Parser_new("sample.v10", input)
	root = parser.parse()
	root_node = root
	sc = Scope_new()
	scope = sc

	v = Constructor_new()
	root.m_visit(v)
	v = Typing_new()
	root.m_visit(v)
	v = Method_new()
	root.m_visit(v)
	v = Definition_new()
	root.m_visit(v)
	v = Resolving_new()
	root.m_visit(v)
	v = IndexCall_new()
	root.m_visit(v)
	v = Printer_new(0, STDOUT, 9999, false)
	root.m_visit(v)
}
