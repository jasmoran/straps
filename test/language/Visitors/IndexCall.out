nBLOCK
  nCLASS(Any)
    nLIST
    nLIST
  nCLASS(Character)
    nLIST
    nLIST
  nCLASS(Number)
    nLIST
    nLIST
  nCLASS(Boolean)
    nLIST
    nLIST
  nENUM(Boolean)
    nBLOCK
      nIDENTIFIER(Unknown: "false")
      nIDENTIFIER(Unknown: "true")
  nCLASS(String)
    nLIST
    nLIST
  nFUNCTION(<Any>: allocate_Number)
    nLIST
      nIDENTIFIER(<Number>: "size")
    nLIST
    nBLOCK
  nFUNCTION(<Boolean>: thisisafunction_Any_Any)
    nLIST
      nIDENTIFIER(<Any>: "arg1")
      nIDENTIFIER(<Any>: "arg2")
    nLIST
      nIDENTIFIER(<Any>: "vrbl1")
      nIDENTIFIER(<Number>: "vrbl2")
    nBLOCK
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r13")
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
        nREGISTER(Unknown: "r13")
      nUNARYOP(Unknown: INCREMENT, right)
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-16")
      nBINARYOP(Unknown: ADD)
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nNUMBER(<Number>: "1337")
      nUNARYOP(Unknown: PUSH, left)
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-16")
      nUNARYOP(Unknown: PUSH, left)
        nNUMBER(<Number>: "77")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r12")
        nBINARYOP(<String>: DOT)
          nNUMBER(<Boolean>: "1")
          nINDEX(<String>: "0")
      nUNARYOP(Unknown: PUSH, left)
        nBINARYOP(<Character>: DOT)
          nREGISTER(Unknown: "r12")
          nINDEX(<Character>: "0")
      nBINARYOP(<String>: PAREN_OPEN)
        nFUNCTION(<String>: _Character_toString_Character)
          nLIST
            nIDENTIFIER(<Character>: "this")
          nLIST
          nBLOCK
        nCALL
          nBINARYOP(<Character>: DOT)
            nREGISTER(Unknown: "r12")
            nINDEX(<Character>: "0")
      nUNARYOP(Unknown: PUSH, left)
        nREGISTER(Unknown: "rax")
      nBINARYOP(<Any>: PAREN_OPEN)
        nFUNCTION(<Any>: _String_println_String_Number_Number)
          nLIST
            nIDENTIFIER(<String>: "this")
            nIDENTIFIER(<Number>: "stream")
            nIDENTIFIER(<Number>: "ilikenumbers")
          nLIST
          nBLOCK
        nCALL
          nREGISTER(Unknown: "rax")
          nNUMBER(<Number>: "77")
          nBINARYOP(<Number>: SQUARE_OPEN)
            nREGISTER(Unknown: "rbp")
            nNUMBER(#Number#: "-16")
      nBINARYOP(<Any>: DOT)
        nREGISTER(Unknown: "rax")
        nINDEX(<Any>: "0")
      nLOOP
        nBLOCK
          nUNARYOP(Unknown: PUSH, left)
            nNUMBER(<Number>: "2")
          nUNARYOP(Unknown: PUSH, left)
            nNUMBER(<Number>: "1")
          nUNARYOP(Unknown: PUSH, left)
            nSTRING(<String>: "oopsy daisy")
          nBINARYOP(<Any>: PAREN_OPEN)
            nFUNCTION(<Any>: _String_println_String_Number_Number)
              nLIST
                nIDENTIFIER(<String>: "this")
                nIDENTIFIER(<Number>: "stream")
                nIDENTIFIER(<Number>: "ilikenumbers")
              nLIST
              nBLOCK
            nCALL
              nSTRING(<String>: "oopsy daisy")
              nNUMBER(<Number>: "1")
              nNUMBER(<Number>: "2")
          nBLOCK
      nRETURN
        nNUMBER(<Boolean>: "1")
  nFUNCTION(<Any>: Any_new_Any)
    nLIST
      nIDENTIFIER(<Any>: "deref")
    nLIST
      nIDENTIFIER(<Any>: "obj")
    nBLOCK
      nUNARYOP(Unknown: PUSH, left)
        nNUMBER(<Number>: "8")
      nBINARYOP(<Any>: PAREN_OPEN)
        nFUNCTION(<Any>: allocate_Number)
          nLIST
            nIDENTIFIER(<Number>: "size")
          nLIST
          nBLOCK
        nCALL
          nNUMBER(<Number>: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nREGISTER(Unknown: "rax")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r12")
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r13")
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Any>: DOT)
          nREGISTER(Unknown: "r12")
          nINDEX(<Any>: "0")
        nREGISTER(Unknown: "r13")
      nRETURN
        nBINARYOP(<Any>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<Character>: Character_new)
    nLIST
    nLIST
      nIDENTIFIER(<Character>: "obj")
    nBLOCK
      nUNARYOP(Unknown: PUSH, left)
        nNUMBER(<Number>: "0")
      nBINARYOP(<Any>: PAREN_OPEN)
        nFUNCTION(<Any>: allocate_Number)
          nLIST
            nIDENTIFIER(<Number>: "size")
          nLIST
          nBLOCK
        nCALL
          nNUMBER(<Number>: "0")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Character>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nREGISTER(Unknown: "rax")
      nRETURN
        nBINARYOP(<Character>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<Number>: Number_new)
    nLIST
    nLIST
      nIDENTIFIER(<Number>: "obj")
    nBLOCK
      nUNARYOP(Unknown: PUSH, left)
        nNUMBER(<Number>: "0")
      nBINARYOP(<Any>: PAREN_OPEN)
        nFUNCTION(<Any>: allocate_Number)
          nLIST
            nIDENTIFIER(<Number>: "size")
          nLIST
          nBLOCK
        nCALL
          nNUMBER(<Number>: "0")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nREGISTER(Unknown: "rax")
      nRETURN
        nBINARYOP(<Number>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<Boolean>: Boolean_new_String)
    nLIST
      nIDENTIFIER(<String>: "as_string")
    nLIST
      nIDENTIFIER(<Boolean>: "obj")
    nBLOCK
      nUNARYOP(Unknown: PUSH, left)
        nNUMBER(<Number>: "8")
      nBINARYOP(<Any>: PAREN_OPEN)
        nFUNCTION(<Any>: allocate_Number)
          nLIST
            nIDENTIFIER(<Number>: "size")
          nLIST
          nBLOCK
        nCALL
          nNUMBER(<Number>: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Boolean>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nREGISTER(Unknown: "rax")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r12")
        nBINARYOP(<Boolean>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r13")
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<String>: DOT)
          nREGISTER(Unknown: "r12")
          nINDEX(<String>: "0")
        nREGISTER(Unknown: "r13")
      nRETURN
        nBINARYOP(<Boolean>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<String>: String_new_Character)
    nLIST
      nIDENTIFIER(<Character>: "random_char")
    nLIST
      nIDENTIFIER(<String>: "obj")
    nBLOCK
      nUNARYOP(Unknown: PUSH, left)
        nNUMBER(<Number>: "8")
      nBINARYOP(<Any>: PAREN_OPEN)
        nFUNCTION(<Any>: allocate_Number)
          nLIST
            nIDENTIFIER(<Number>: "size")
          nLIST
          nBLOCK
        nCALL
          nNUMBER(<Number>: "8")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
        nREGISTER(Unknown: "rax")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r12")
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
      nBINARYOP(Unknown: ASSIGN)
        nREGISTER(Unknown: "r13")
        nBINARYOP(<Character>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "16")
      nBINARYOP(Unknown: ASSIGN)
        nBINARYOP(<Character>: DOT)
          nREGISTER(Unknown: "r12")
          nINDEX(<Character>: "0")
        nREGISTER(Unknown: "r13")
      nRETURN
        nBINARYOP(<String>: SQUARE_OPEN)
          nREGISTER(Unknown: "rbp")
          nNUMBER(#Number#: "-8")
  nFUNCTION(<String>: _Character_toString_Character)
    nLIST
      nIDENTIFIER(<Character>: "this")
    nLIST
    nBLOCK
  nFUNCTION(<Any>: _String_println_String_Number_Number)
    nLIST
      nIDENTIFIER(<String>: "this")
      nIDENTIFIER(<Number>: "stream")
      nIDENTIFIER(<Number>: "ilikenumbers")
    nLIST
    nBLOCK
